testpb/1.pulsar.go

```diff
@@ -0,0 +1,2262 @@
+//
+// Clearing an extension field clears both the extension type and value
+// associated with the given field number.
+//
+// Clear is a mutating operation and unsafe for concurrent use.
+func (x A) Clear(descriptor protoreflect.FieldDescriptor) {
+	x.ProtoReflect().Clear(descriptor)
+}
+
+type _A_19_list struct {
+	list []*B
+}
+
+var _ protoreflect.List = (*_A_19_list)(nil)
+
+func (x *_A_19_list) Len() int {
+	return len(x.list)
+}
+
+func (x *_A_19_list) Get(i int) protoreflect.Value {
+	return protoreflect.ValueOfMessage(x.list[i].ProtoReflect())
+}
+
+func (x *_A_19_list) Set(i int, value protoreflect.Value) {
+	unwrapped := value.Message()
+	concreteValue := unwrapped.Interface().(*B)
+	x.list[i] = concreteValue
+}
+
+func (x *_A_19_list) Append(value protoreflect.Value) {
+	unwrapped := value.Message()
+	concreteValue := unwrapped.Interface().(*B)
+	x.list = append(x.list, concreteValue)
+}
+
+func (x *_A_19_list) AppendMutable() protoreflect.Value {
+	v := new(B)
+	x.list = append(x.list, v)
+	return protoreflect.ValueOfMessage(v.ProtoReflect())
+}
+
+func (x *_A_19_list) Truncate(n int) {
+	for i := n; i < len(x.list); i++ {
+		x.list[i] = nil
+	}
+	x.list = x.list[:n]
+}
+
+func (x *_A_19_list) NewElement() protoreflect.Value {
+	v := new(B)
+	x.list = append(x.list, v)
+	return protoreflect.ValueOfMessage(v.ProtoReflect())
+}
+
+func (x *_A_19_list) IsValid() bool {
+	return x.list == nil || len(x.list) == 0
+}
+
+type _A_22_list struct {
+	list []Enumeration
+}
+
+var _ protoreflect.List = (*_A_22_list)(nil)
+
+func (x *_A_22_list) Len() int {
+	return len(x.list)
+}
+
+func (x *_A_22_list) Get(i int) protoreflect.Value {
+	return protoreflect.ValueOfEnum((protoreflect.EnumNumber)(x.list[i]))
+}
+
+func (x *_A_22_list) Set(i int, value protoreflect.Value) {
+	unwrapped := value.Enum()
+	concreteValue := (Enumeration)(unwrapped)
+	x.list[i] = concreteValue
+}
+
+func (x *_A_22_list) Append(value protoreflect.Value) {
+	unwrapped := value.Enum()
+	concreteValue := (Enumeration)(unwrapped)
+	x.list = append(x.list, concreteValue)
+}
+
+func (x *_A_22_list) AppendMutable() protoreflect.Value {
+	panic(fmt.Errorf("AppendMutable can not be called on message A at list field LIST_ENUM as it is not of Message kind"))
+}
+
+func (x *_A_22_list) Truncate(n int) {
+	x.list = x.list[:n]
+}
+
+func (x *_A_22_list) NewElement() protoreflect.Value {
+	v := 0
+	return protoreflect.ValueOfEnum((protoreflect.EnumNumber)(v))
+}
+
+func (x *_A_22_list) IsValid() bool {
+	return x.list == nil || len(x.list) == 0
+}
+
+// Get retrieves the value for a field.
+//
+// For unpopulated scalars, it returns the default value, where
+// the default value of a bytes scalar is guaranteed to be a copy.
+// For unpopulated composite types, it returns an empty, read-only view
+// of the value; to obtain a mutable reference, use Mutable.
+func (x *A) Get(descriptor protoreflect.FieldDescriptor) protoreflect.Value {
+	switch descriptor.Name() {
```

Comment:
> What if we just used something like the dynamicList from dynamicpb?

I've tried combining dynamicpb + normal proto messages and it ends up in invalid type panics -> `invalid type: got *dynamicpb.Message, want *v1alpha1.Coin [recovered]` - `dynamicpb` also does not give us access to dynamiclist constructors.

we could create a "wrapper", I was experimenting with it but then i got stuck at some point during `message.Set` where field is a list field, and value is a `protoreflect.ListValue` in which you'd need to cast the value to the slice type.

Reply:
> Why not switching on field number? Would that maybe be faster?

The Get logic will need to be rewritten partially (to also include extension support),  because it's not that safe, meaning that using just field name or number I think might not be sufficient, probably the safest way (assuming we're working with a global protobuf registry) is to use the field fullname. Reason being to avoid for example having a field which is `number: 1, kind: message` being set using a field `number: 1, kind: int64`.

The user of `Get` would be left with an invalid `protoreflect.Value`, and the panic would happen only when `value` is being unwrapped.